source/server/connection_handler_impl.cc

```diff
@@ -44,19 +45,24 @@ void ConnectionHandlerImpl::addListener(absl::optional<uint64_t> overridden_list
     if (overridden_listener.has_value()) {
       for (auto& listener : listeners_) {
         if (listener.second.listener_->listenerTag() == overridden_listener) {
-          listener.second.tcp_listener_->get().updateListenerConfig(config);
+          listener.second.tcpListener()->get().updateListenerConfig(config);
           return;
         }
       }
       NOT_REACHED_GCOVR_EXCL_LINE;
     }
     auto tcp_listener = std::make_unique<ActiveTcpListener>(*this, config);
-    details.tcp_listener_ = *tcp_listener;
+    details.typed_listener_ = *tcp_listener;
     details.listener_ = std::move(tcp_listener);
   } else {
     ASSERT(config.udpListenerFactory() != nullptr, "UDP listener factory is not initialized.");
-    details.listener_ =
-        config.udpListenerFactory()->createActiveUdpListener(*this, dispatcher_, config);
+    ASSERT(worker_index_.has_value());
+    details.listener_ = config.udpListenerFactory()->createActiveUdpListener(*worker_index_, *this,
+                                                                             dispatcher_, config);
+    Network::UdpListenerCallbacks* udp_listener =
+        dynamic_cast<ActiveUdpListenerBase*>(details.listener_.get());
```

Comment:
Why is the dynamic_cast needed here? This seems bad? Can you add a TODO of some type to figure out how to clean this up?

IIRC there is some mess here that we have dealt with in the past around extensions actually implementing active listener base, etc.? Is that right? It would be good to have a comment if that is the case. I know I have commented on similar things in the past. I think possible if we make QUIC/QUICHE not extension based this gets better.

Reply:
Good point.

I fixed this by promoting part of ActiveUdpListenerBase to /include, so that the factory can return that type. No more casts needed. Take a look at the commit; I had to switch inheritance of ActiveListener to virtual. If we don't like that, the same can be accomplished by making ActiveListenerImplBase into a mixin template, or templatizing whether it inherits from ActiveListener or ActiveUdpListener. Either is fine with me.